### ISSUE TITLE
Performance problem when compiling with -O0

### ISSUE THREAD

**Issue by Voultapher** [ericniebler/range-v3#458]
There is a strong performance implication when compiling unoptimized.

I tried making the example simple and representative.

Here is an example where using ranges resulted in a 10x performance impact on windows 10 using Clang 3.7 on an i7-5500U

	#include <iostream>
	#include <vector>
	#include <chrono>
	#include <algorithm>
	#include <random>

	#include <range\v3\all.hpp>

	template<typename K> auto createKeys(const size_t size, K startKey)
	{
		std::vector<K> keyVals;

		std::mt19937_64 generator(1337); // fixed seed for reproducibility
		std::uniform_int_distribution<int64_t> distribution(0, size / 10);
		std::vector<K> usedRandValues;
		K randVal;

		for (size_t i = 0; i < size; ++i)
		{
			keyVals.push_back(startKey);
			randVal = distribution(generator);
			startKey = randVal;
		}

		return keyVals;
	}

	template<typename T> std::vector<T> sortReplace(std::vector<T> vec)
	{
		std::sort(vec.begin(), vec.end());

		auto last = std::unique(vec.begin(), vec.end());
		vec.erase(last, vec.end());

		return vec;
	}

	template<typename T> std::vector<T> copy_sort_unique(std::vector<T> vec)
	{
		return std::move(vec) | ranges::action::sort | ranges::action::unique;
	}

	template<typename T, typename FUNC> int64_t benchmark(
		const size_t testSize,
		FUNC func	
	)
	{
		size_t repeatCount = 5; // to reduce statistical variance
		auto keyVals = createKeys(testSize, T());

		auto start = std::chrono::high_resolution_clock::now();
		for (size_t i = 0; i < repeatCount; ++i)
			func(keyVals);
		auto end = std::chrono::high_resolution_clock::now();

		return std::chrono::duration_cast<std::chrono::nanoseconds>
			(end - start).count() / repeatCount;
	}

	int main()
	{
		size_t testSize = 2000;
		using val_t = double;

		auto sortReplaceResult = benchmark<val_t>(testSize, sortReplace<val_t>);
		auto rangesResult = benchmark<val_t>(testSize, copy_sort_unique<val_t>);

		std::cout << sortReplaceResult << "\n";
		std::cout << rangesResult << "\n";

		std::cin.get();
		return 0;
	}

**Comment by CaseyCarter** [comment]
> Here is an example where using ranges resulted in a 10x performance impact:

What OS/compiler/standard library/build commands?

**Comment by CaseyCarter** [comment]
Having actually examined your sample program, I doubt the performance difference you're seeing has anything to do with ranges. Specifically, the only use of ranges is in `copy_sort_unique`:
```c++
template<typename T> std::vector<T> copy_sort_unique(std::vector<T> vec)
{
    return std::move(vec) | ranges::action::sort | ranges::action::unique;
}
```
`std::move(vec) | ranges::action::sort | ranges::action::unique` is effectively equivalent to:
```c++
ranges::sort(vec);
vec.erase(ranges::unique(vec), ranges::end(vec));
```
which greatly resembles your non-ranges code, except that it uses range-v3's implementations of the `sort` and `unique` algorithms instead of those in the standard library. I think you'll find that *this* is the contributing factor to the observed performance differences: many standard library implementations have iterators that are hostile to performance in debug builds, and convert those iterators into less crappy iterators at algorithm boundaries to improve performance.

For example, if I change `sortReplace` to:
```c++
template<typename T> std::vector<T> sortReplace(std::vector<T> vec)
{
#if USE_PTR
    auto first = vec.data();
    auto last = first + vec.size();
#else
    auto first = vec.begin();
    auto last = vec.end();
#endif

#if USE_STL
    std::sort(first, last);
    last = std::unique(first, last);
#else
    ranges::sort(first, last);
    last = ranges::unique(first, last);
#endif

    vec.erase(vec.begin() + (last - first), vec.end());
    return vec;
}
```
and then compile (Windows 10, using the Clang/C2 compiler from VS2015 Update 3):
<pre>
Casey@KIRITO C:\Users\Casey\Desktop
> "c:\Program Files (x86)\Microsoft Visual Studio 14.0\vc\vcvarsall.bat" x86

Casey@KIRITO C:\Users\Casey\Desktop
> "c:\Program Files (x86)\Microsoft Visual Studio 14.0\vc\ClangC2\bin\x86\clang.exe" -fexceptions -I D:\dev\range-v3\include -O0 -DUSE_STL=0 -DUSE_PTR=0 test.cpp && .\a.exe
1219266
1203280

Casey@KIRITO C:\Users\Casey\Desktop
> "c:\Program Files (x86)\Microsoft Visual Studio 14.0\vc\ClangC2\bin\x86\clang.exe" -fexceptions -I D:\dev\range-v3\include -O0 -DUSE_STL=0 -DUSE_PTR=1 test.cpp && .\a.exe
584160
1156751

Casey@KIRITO C:\Users\Casey\Desktop
> "c:\Program Files (x86)\Microsoft Visual Studio 14.0\vc\ClangC2\bin\x86\clang.exe" -fexceptions -I D:\dev\range-v3\include -O0 -DUSE_STL=1 -DUSE_PTR=0 test.cpp && .\a.exe
346664
1165895

Casey@KIRITO C:\Users\Casey\Desktop
> "c:\Program Files (x86)\Microsoft Visual Studio 14.0\vc\ClangC2\bin\x86\clang.exe" -fexceptions -I D:\dev\range-v3\include -O0 -DUSE_STL=1 -DUSE_PTR=1 test.cpp && .\a.exe
327070
1249311
</pre>
Notice that `USE_PTR=0` (using the library's vector iterators) in both `USE_STL=0/1` cases has different performance than `USE_PTR=1` - this tells us immediately that the standard library doesn't use pointers for its vector iterators. 

The `USE_STL=0, USE_PTR=0` case has roughly the same performance for the range and non-range cases as expected. (Well, as expected by me, anyway ;)

The `USE_STL=0, USE_PTR=1` gets a 100% speedup by avoiding the non-pointer iterators. Comparing with `USE_STL=1, USE_PTR=1` shows that the standard library's implementation of `sort` is much faster than that in range-v3 (See #406 "`ranges::sort` is slow"), 

The performance delta from `USE_STL=1, USE_PTR=0` to `USE_STL=0, USE_PTR=0` is roughly 3x, which differs substantially from the roughly 1.8x delta from `USE_STL=1, USE_PTR=1` to `USE_STL=0, USE_PTR=1`. This indicates to me that the standard library is using its full knowledge of the iterator type to convert to pointers before sorting.

In short - if your platform is similar to mine - the performance problem here isn't the use of ranges so much as that range-v3 doesn't know as much about your standard library's iterators as do the maintainers of that standard library. If we want to eliminate this performance delta we would need to detect sets of parameters to algorithms that are compatible with the standard library algorithm specifications, and delegate to the standard library for such parameter sets.

**Comment by Voultapher** [comment]
Thanks for the very informative answer! Given that there is already an issue for performance concerning sort, should I close this issue?

**Comment by ericniebler** [comment]
Thanks @CaseyCarter.
